torch/csrc/jit/codegen/cuda/manager.cpp

```diff
@@ -81,26 +82,272 @@ class CudaFusionManager {
+    // `permute_seq` is the returned permutation to achieve sorted stride;
+    at::DimVector permute_seq;
+
+    auto stride_properties = type->stride_properties().sizes();
+
+    TORCH_INTERNAL_ASSERT(
+        stride_properties.has_value(),
+        "unknown sizes or stride_properties, collapsing shouldn't happen");
+
+    // TODO: reuse this;
+    const int rank = static_cast<int>(stride_properties->size());
+
+    // stores axes with stride_index;
+    std::set<int> ordered_axes;
+
+    // TODO: this does not support broadcast yet;
+    for (int i = 0; i < rank; i++) {
+      if (auto index = (*stride_properties)[i]->stride_index_) {
+        ordered_axes.insert(*index);
+      }
+    }
+
+    int unallocated_axis = 0;
+    // we push from slowest to fastest
+    for (int i = rank - 1; i >= 0; i--) {
+      if (auto index = (*stride_properties)[i]->stride_index_) {
+        // pushing axis index to current entry in permute_seq;
+        permute_seq.emplace_back(*index);
+      } else {
+        // no designated axis for this slot, so we push an axis w/o designated
+        // order;
+        while (ordered_axes.count(unallocated_axis) != 0) {
+          ++unallocated_axis;
+        }
+        permute_seq.emplace_back(unallocated_axis++);
+      }
+    }
+    return permute_seq;
+  }
+
+  std::vector<IValue> dimSortInputs(
+      std::shared_ptr<Graph>& graph,
+      const at::ArrayRef<IValue> inputs) {
+    if (!IsNewExecutorEnabled() || graphHasReduction(graph)) {
+      return inputs.vec();
+    }
+    auto acc_type = mergeInputTensorType(graph);
+
+    if (!acc_type->dim().has_value()) {
+      return inputs.vec();
+    }
+
+    auto strategy = getSortStrideScheme(acc_type);
+    // TODO: early return if permutation is no-op;
+
+    std::vector<IValue> permuted_inputs;
+    for (const auto& input : inputs) {
+      if (input.isTensor()) {
+        permuted_inputs.emplace_back(input.toTensor().permute(strategy));
+      } else {
+        permuted_inputs.emplace_back(input);
+      }
+    }
+    return permuted_inputs;
+  }
+
+  std::vector<at::Tensor> dimSortOutputs(
+      const std::shared_ptr<Graph>& graph,
+      const std::vector<at::Tensor>& outputs) {
+    if (!IsNewExecutorEnabled() || graphHasReduction(graph)) {
+      return outputs;
+    }
+    auto acc_type = mergeInputTensorType(graph);
+    if (!acc_type->dim().has_value()) {
+      return outputs;
+    }
+
+    auto strategy = getSortStrideScheme(acc_type);
+    // TODO: early return if permutation is no-op;
+    auto restore_strategy = restorePermutation(strategy);
+
+    std::vector<at::Tensor> permuted_outputs;
+    TORCH_INTERNAL_ASSERT(outputs.size() == graph->outputs().size());
+    permuted_outputs.reserve(outputs.size());
+    for (const auto& output : outputs) {
+      permuted_outputs.emplace_back(output.permute(restore_strategy));
+    }
+    return permuted_outputs;
+  }
+
+  // two thing need to be adjusted:
```

Comment:
Does this support reduction? Would assume there'd be similar issues with sorting.

Reply:
Same restriction applies and we are excluding graph with reduction operation for now.